### ISSUE TITLE
Reproducing log loss with custom objective

### ISSUE THREAD

**Issue by MaxHalford** [microsoft/LightGBM#3312]
Hello. 

I'm having some issues reproducing the default objective function for classification. I'm using the `loglikelihood` definition from the [Python advanced example](https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/advanced_example.py). I've attached a reproducible example.

First I use the credit cards dataset:

```sh
$ wget maxhalford.github.io/files/datasets/creditcardfraud.zip
$ unzip creditcardfraud.zip
```

Then, I load and split my dataset with `pandas`:

```py
import pandas as pd
from sklearn import model_selection

df = pd.read_csv('/Users/mhalford/creme_data/CreditCard/creditcard.csv')
X = df.drop(columns='Class')
y = df['Class']

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, random_state=42)
X_fit, X_val, y_fit, y_val = model_selection.train_test_split(X_train, y_train, random_state=42)
```

The following snippet uses `LightGBM` with default parameters, which as far as I can tell uses logarithmic loss:

```py
import lightgbm
from sklearn import metrics

fit = lightgbm.Dataset(X_fit, y_fit)
val = lightgbm.Dataset(X_val, y_val)

model = lightgbm.train(
    params={
        'learning_rate': .01,
        'objective': 'binary',
        'metric': 'binary_logloss',
    },
    train_set=fit,
    num_boost_round=10_000,
    valid_sets=(fit, val),
    valid_names=('fit', 'val'),
    early_stopping_rounds=20,
    verbose_eval=100,
)

y_pred = model.predict(X_test)

print()
print(f'ROC AUC: {metrics.roc_auc_score(y_test, y_pred):.5f}')
print(f'Log loss: {metrics.log_loss(y_test, y_pred):.5f}')
```

This gives me the following output:

```
Training until validation scores don't improve for 20 rounds
[100]	fit's binary_logloss: 0.00191083	val's binary_logloss: 0.00358371
[200]	fit's binary_logloss: 0.000825181	val's binary_logloss: 0.00286873
[300]	fit's binary_logloss: 0.000403679	val's binary_logloss: 0.00262094
Early stopping, best iteration is:
[355]	fit's binary_logloss: 0.000282887	val's binary_logloss: 0.00257033

ROC AUC: 0.97721
Log loss: 0.00233
```

So far, so good. Now when I use a custom objective function that implements the logarithmic loss, I don't get the same results. I'm aware that I have to modify my metric to take into the fact that `preds` contains the margins and not the probabilities. I therefore transform the margins with the sigmoid function before computing my objective and my metric.

```py
def sigmoid(x):
    return 1. / (1. + np.exp(-x))


def log_loss_objective(preds, train_data):
    labels = train_data.get_label()
    preds = sigmoid(preds)
    grad = preds - labels
    hess = preds * (1. - preds)
    return grad, hess


def log_loss_metric(preds, train_data):
    labels = train_data.get_label()
    preds = sigmoid(preds)
    return 'error', metrics.log_loss(labels, preds), False


model = lightgbm.train(
    params={
        'learning_rate': .01
    },
    train_set=fit,
    num_boost_round=10_000,
    valid_sets=(fit, val),
    valid_names=('fit', 'val'),
    early_stopping_rounds=20,
    verbose_eval=100,
    
    fobj=log_loss_objective,
    feval=log_loss_metric,
)

y_pred = sigmoid(model.predict(X_test))

print()
print(f'ROC AUC: {metrics.roc_auc_score(y_test, y_pred):.5f}')
print(f'Log loss: {metrics.log_loss(y_test, y_pred):.5f}')
```

This gives me the following output, which is different from above:

```
Training until validation scores don't improve for 20 rounds
[100]	fit's error: 0.203631	val's error: 0.203803
[200]	fit's error: 0.0710043	val's error: 0.0712822
[300]	fit's error: 0.0263409	val's error: 0.0267795
[400]	fit's error: 0.0103281	val's error: 0.011038
[500]	fit's error: 0.00422448	val's error: 0.00539362
[600]	fit's error: 0.00181789	val's error: 0.00339196
[700]	fit's error: 0.000789874	val's error: 0.00273831
[800]	fit's error: 0.00038473	val's error: 0.00256254
Early stopping, best iteration is:
[833]	fit's error: 0.000311434	val's error: 0.00254163

ROC AUC: 0.97189
Log loss: 0.00222
```

At least the outputs are somewhat similar in both cases, but I would be expecting exactly the same outputs. I think it would be to nice to able to reproduce the default behaviour with a custom objective function, as this is the basis for building fancier objectives. Indeed, I don't feel comfortable implementing a new objective knowing that I'm not able to reproduce the default behaviour.  

Could you please me out to something I'm missing? I've checked on StackOverflow, Kaggle, and GitHub and can't find an answer...

Kindest regards.

**Comment by shiyu1994** [comment]
This is just because the parameter settings are 'inconsistent' in two snippets. In the `params` of your first snippet, set `boost_from_average: False`. Then you will get exactly the same result as using your customized log loss function. 

By default, `boost_from_average` is `True`, which means LightGBM will adjust initial scores of all data points to the mean of labels for faster convergence. Note that here 'mean' of labels is calculated differently for different objective functions. For customized objective function, it is unclear how to calculate this 'mean', so 'boost_from_average' is actually disabled. If you want to boost from a specific score, you can set the init scores for the datasets.

For more details about the init score of `boost_from_average` in log loss case, you may refer to the following code

> https://github.com/microsoft/LightGBM/blob/e9fbd19d7cbaeaea1ca54a091b160868fc5c79ec/src/objective/binary_objective.hpp#L135-L158

**Comment by MaxHalford** [comment]
Hey @shiyu1994, thanks for telling me about the initialization parameters. I wouldn't have found it by myself. I've looked around and it seems that the goal is to find the constant that minimizes the loss function over the ground truths. Cheers!

**Comment by MaxHalford** [comment]
Sorry I'm reopening this because I have a final question, I closed it too hastily.

I've taken into account your help and I've set the `init_score` parameter of both data datasets, as so:

```py
import numpy as np


def log_loss_init_score(y):
    pavg = y.mean()
    pavg = np.clip(pavg, 1e-15, 1 - 1e-15)
    g = np.log(pavg / (1. - pavg))
    return np.full_like(y, g, dtype=float)


fit = lightgbm.Dataset(X_fit, y_fit, init_score=log_loss_init_score(y_fit))
val = lightgbm.Dataset(X_val, y_val, init_score=log_loss_init_score(y_val), reference=fit)


def sigmoid(x):
    return 1 / (1 + np.exp(-x))


def log_loss_objective(preds, train_data):
    labels = train_data.get_label()
    preds = sigmoid(preds)
    grad = preds - labels
    hess = preds * (1 - preds)
    return grad, hess


def log_loss_metric(preds, train_data):
    labels = train_data.get_label()
    preds = sigmoid(preds)
    return 'error', metrics.log_loss(labels, preds), False


model = lightgbm.train(
    params={
        'learning_rate': .01,
        'boost_from_average': True
    },
    train_set=fit,
    num_boost_round=10_000,
    valid_sets=(fit, val),
    valid_names=('fit', 'val'),
    early_stopping_rounds=20,
    verbose_eval=100,
    
    fobj=log_loss_objective,
    feval=log_loss_metric,
)

y_pred = sigmoid(model.predict(X_test))

print()
print(f'ROC AUC: {metrics.roc_auc_score(y_test, y_pred):.5f}')
print(f'Log loss: {metrics.log_loss(y_test, y_pred):.5f}')
```

This gives me the following output, which is what I'm expecting (there's some very slight difference that I assume is down to numerical precision). However, the log loss is not what I'm expecting, even though I've converted the margins to probabilities using the sigmoid function.

```
Training until validation scores don't improve for 20 rounds
[100]	fit's error: 0.00191083	val's error: 0.00358463
[200]	fit's error: 0.000825181	val's error: 0.00286665
[300]	fit's error: 0.000403679	val's error: 0.00261748
Early stopping, best iteration is:
[355]	fit's error: 0.000282887	val's error: 0.00256643

ROC AUC: 0.97721
Log loss: 0.05873
```

Can you see the last piece of the puzzle that I can't?

**Comment by MaxHalford** [comment]
Actually, I just found out what's missing: the initialization score used in the training set needs to be added to the margins produced for the test set, as so:

```py
import numpy as np


def log_loss_init_score(y):
    pavg = y.mean()
    pavg = np.clip(pavg, 1e-15, 1 - 1e-15)
    g = np.log(pavg / (1. - pavg))
    return g


fit = lightgbm.Dataset(X_fit, y_fit, init_score=np.full_like(y_fit, log_loss_init_score(y_fit), dtype=float))
val = lightgbm.Dataset(X_val, y_val, init_score=np.full_like(y_val, log_loss_init_score(y_fit), dtype=float), reference=fit)


def sigmoid(x):
    return 1 / (1 + np.exp(-x))


def log_loss_objective(preds, train_data):
    labels = train_data.get_label()
    preds = sigmoid(preds)
    grad = preds - labels
    hess = preds * (1 - preds)
    return grad, hess


def log_loss_metric(preds, train_data):
    labels = train_data.get_label()
    preds = sigmoid(preds)
    return 'error', metrics.log_loss(labels, preds), False


model = lightgbm.train(
    params={
        'learning_rate': .01,
        'boost_from_average': True
    },
    train_set=fit,
    num_boost_round=10_000,
    valid_sets=(fit, val),
    valid_names=('fit', 'val'),
    early_stopping_rounds=20,
    verbose_eval=100,
    
    fobj=log_loss_objective,
    feval=log_loss_metric,
)

y_pred = sigmoid(log_loss_init_score(y_fit) + model.predict(X_test))

print()
print(f'ROC AUC: {metrics.roc_auc_score(y_test, y_pred):.5f}')
print(f'Log loss: {metrics.log_loss(y_test, y_pred):.5f}')
```

To be honest, I find this very surprising. Don't you think that LightGBM should take the `init_score` of the training set into account if I specify it?

**Comment by shiyu1994** [comment]
The init score is a part of dataset. The model is only responsible for recording values stored in trees. If `boost_from_average` is enabled, the mean score will be a part of the model. However, if the init score is specified through the dataset, the model is not responsible for this since different datasets can have different init scores.

**Comment by MaxHalford** [comment]
Fair enough!

**Comment by HuaYZhao** [comment]
If I add the colsample_bytree parameter, it will become inconsistent again, how do I solve this?
